### ISSUE TITLE
Large precache downloads may result in an MSC error

### ISSUE THREAD

**Issue by aysiu** [munki/munki#897]
For some especially large downloads (e.g., a 6 GB macOS installer) that are precached items, even when the download to the /Library/Managed Installs/Cache is successful, the user may still see this error message pop up in Managed Software Center:
```
Update check failed
There is a configuration problem with the managed software installer. The process ended unexpectedly. Contact your systems administrator
```

**Comment by gregneagle** [comment]
To be more precise, this issue occurs when the precaching downloads do not complete until after the `managedsoftwareupdate` session completes. This could be because of one really large download, or 100 small downloads.

See https://groups.google.com/d/msg/munki-discuss/WlDkz8KoyTQ/J2z74O2xEQAJ

> The code the precaching_agent to download installer items is also used by mangedsoftwareupdate. It posts progress messages (what it's downloading, percentage complete, etc). When MSC.app is open, this confuses it. It thinks managedsoftwareupdate is running and posting status updates, and then when the precaching_agent quits MSC.app thinks managedsoftwareupdate has exited unexpectedly.

**Comment by aysiu** [comment]
Latest Munki build (3.5.2.3633), ran on a client machine (using Managed Software Center's GUI) that has only Mojave's installer as a `precache` optional install (nothing else in the manifest), and it started the download but didn't finish.

```
Jan 08 2019 13:50:24 -0600     #### Beginning precaching session ####
Jan 08 2019 13:50:24 -0600     Downloading Install macOS Mojave-10.14.2.dmg from Install macOS Mojave-10.14.2.dmg
Jan 08 2019 13:50:24 -0600     Downloading Install macOS Mojave-10.14.2.dmg from Install macOS Mojave-10.14.2.dmg
Jan 08 2019 13:50:25 -0600     Downloading Install macOS Mojave-10.14.2.dmg...
Jan 08 2019 13:50:25 -0600 Checking Apple Software Update catalog...
Jan 08 2019 13:50:25 -0600     Caching CatalogURL https://swscan.apple.com/content/catalogs/others/index-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog
Jan 08 2019 13:51:07 -0600 Checking for available Apple Software Updates...
Jan 08 2019 13:51:08 -0600     Finding available software
Jan 08 2019 13:51:38 -0600     No updates are available.
Jan 08 2019 13:51:38 -0600 Finishing...
```

It moved so fast from downloading to the next stage (checking for Apple software updates) that I wondered for a second whether the download was supposed to work asynchronously.

It did get a partial download of 1.22 GB (out of 6 GB): `Install macOS Mojave-10.14.2.dmg.download`

**Comment by aysiu** [comment]
Wait—is it actually downloading asynchronously? I clicked **Check Again**, and in the next Munki run, it deleted the previous partial download and started again, but even after the Munki run was completely finished, when I keep going back to check the /Library/Managed Installs/Cache folder, the download has gotten bigger.

**Comment by aysiu** [comment]
Looks as if it worked!
```
Jan 08 2019 14:00:39 -0600 ### Ending managedsoftwareupdate run ###
Jan 08 2019 14:04:42 -0600 WARNING: Failed to precache the installer for Install_macOS_Mojave because HTTP result 200: no error
Jan 08 2019 14:04:42 -0600     #### Ending precaching session ####
Jan 08 2019 14:06:11 -0600 ### Starting managedsoftwareupdate run: auto ###
```
Even though there's a warning about a failed precache, eventually it did finish downloading. I even ran an `shasum -a 256` on the download, and it matches what it should.

**Comment by gregneagle** [comment]
"Wait—is it actually downloading asynchronously?" -- yes. That is what `/usr/bin/munki/precache_agent` _does_. And that was the cause of the issue here: precache_agent was sending progress update messages that MSC.app was receiving and interpreting as messages from `managedsoftwareupdate`.

**Comment by aysiu** [comment]
That's awesome! I didn't know that's what the agent did. It actually makes sense, too, from just a user experience perspective (which is why I'm sure you created the separate agent)—you don't want to wait to be able to do anything else while there's a download of something you may not even want to install.

**Comment by gregneagle** [comment]
Exactly.

**Comment by gregneagle** [comment]
I'm unhappy about this observation: "I clicked Check Again, and in the next Munki run, it deleted the previous partial download and started again"...

**Comment by aysiu** [comment]
Is it downloading in a different way from Munki's set up to download other stuff? I've always found Munki to continue downloading interrupted downloads.

Is there someone else who can verify this behavior? I'll certainly try it again myself, but it'd be nice to have more testers.

To replicate:
1. Make a new `precache` optional install, preferably one that's several GB large
2. Launch up Managed Software Center
3. Once the Munki run is complete (but the precache agent is clearly still running), click Check Again in MSC and see if the incomplete download deletes before re-downloading

I'll try it again myself, but... yeah, if someone else (not necessarily Greg) can test, that'd be good, too.

**Comment by aysiu** [comment]
I enabled logging level 3 in the ManagedSoftwareUpdate.log, and I didn't see anything indicating what happened when the partial downlaod got deleted.

I did notice that at some point the .download extension on the partial download turns into a .dmg extention, and it _looks_ like a completed download (even though it's not)—somewhere between 1.5 GB and 6 GB.

Once it gets the .dmg extension, clicking **Check Again** does not delete the partial download.
